clang/lib/CodeGen/CGExprCXX.cpp

```diff
@@ -1038,11 +1038,14 @@ void CodeGenFunction::EmitNewArrayInitializer(
     return true;
   };
 
+  const InitListExpr *ILE = dyn_cast<InitListExpr>(Init);
+  const CXXParenListInitExpr *CPLIE = dyn_cast<CXXParenListInitExpr>(Init);
+  const StringLiteral *SL = dyn_cast<StringLiteral>(Init);
   // If the initializer is an initializer list, first do the explicit elements.
-  if (const InitListExpr *ILE = dyn_cast<InitListExpr>(Init)) {
+  if (ILE || CPLIE || SL) {
```

Comment:
This code is assuming the initializer for an array must be one of the following:

- An InitListExpr with a list of elements
- A CXXParentListInitExpr with a list of elements
- An InitListExpr with a "string" (isStringLiteralInit is true)
- A StringLiteral

Looking at the standard, I guess this is a complete list.

The last one seems a little suspect to me, though; if it's using the same IsStringInit() as InitListExpr does, a "string" might not be a StringLiteral.

Reply:
This is parsed as a `StringLiteral` as shown in the AST.

`new-array.cc`:

```cpp
char* x = new char[]("asdf");
```

```
$ ./bin/clang -Xclang -ast-dump -std=c++20 ~/src/tests/new-array.cc

(irrelevant bits omitted)

|-VarDecl 0x557ce6fee8c0 </usr/local/google/home/ayzhao/src/tests/new-array.cc:1:1, col:28> col:7 x 'char *' cinit
| `-CXXNewExpr 0x557ce700cca0 <col:11, col:28> 'char *' array Function 0x557ce6fef0a0 'operator new[]' 'void *(unsigned long)'
|   |-IntegerLiteral 0x557ce700cc50 <col:15> 'unsigned long' 5
|   `-StringLiteral 0x557ce6feea00 <col:22> 'char[5]' "asdf"
```

Reply:
Sure, but have you tried, for example, `char* x = new char[](__extension__ "asdf");`.  Or in objective-c++ mode, `char* x = new char[](@encode(int));`

Reply:
Thanks again for catching these - they both cause crashes.

For the `__extension__` example, we have a `StringLiteral` wrapped in a `UnaryOperator`

```
|-VarDecl 0x561d5f1ba190 </usr/local/google/home/ayzhao/src/tests/extension.cc:1:1, col:42> col:7 x 'char *' cinit
| `-CXXNewExpr 0x561d5f1d8580 <col:11, col:42> 'char *' array Function 0x561d5f1ba990 'operator new[]' 'void *(unsigned long)'
|   |-IntegerLiteral 0x561d5f1d8530 <col:15> 'unsigned long' 5
|   `-UnaryOperator 0x561d5f1ba2f0 <col:22, col:36> 'char[5]' prefix '__extension__' cannot overflow
|     `-StringLiteral 0x561d5f1ba2d0 <col:36> 'char[5]' "asdf"
```

and for the Objective-C++ example, we have a `ObjCEncodeExpr`:

```
|-VarDecl 0x55b8a4d0a150 </usr/local/google/home/ayzhao/src/tests/encode.mii:1:1, col:34> col:7 x 'char *' cinit
| `-CXXNewExpr 0x55b8a4d2a560 <col:11, col:34> 'char *' array Function 0x55b8a4d0a940 'operator new[]' 'void *(unsigned long)'
|   |-IntegerLiteral 0x55b8a4d2a510 <col:15> 'unsigned long' 2
|   `-ObjCEncodeExpr 0x55b8a4d0a2a0 <col:22, col:33> 'char[2]' 'int'
```

Reply:
OK, I've implemented a fix for this locally (but I haven't pushed my branch yet as I haven't written tests).

I'm running into a problem with the example:

```objc
char* x = new char[](@encode(int));
```

Clang produces the error:

```
$ bin/clang++ -c -o - -S -emit-llvm -std=c++20 ~/src/tests/encode.mii
/usr/local/google/home/ayzhao/src/tests/encode.mii:1:28: error: cannot compile this aggregate expression yet
    1 | const char* x = new char[](@encode(int));
      |                            ^~~~~~~~~~~~
1 error generated.
```

Clang emits the same error for a braced initialization-list:

```
$ bin/clang++ -c -o - -S -emit-llvm -std=c++20 ~/src/tests/encode.mii
/usr/local/google/home/ayzhao/src/tests/encode.mii:1:28: error: cannot compile this aggregate expression yet
    1 | const char* x = new char[]{@encode(int)};
      |                            ^~~~~~~~~~~~
1 error generated.
```

while GCC happily compiles both examples: https://godbolt.org/z/494x4ne5o

Given that braced initialization of a `ObjCEncodeExpr` is already broken in clang, I'm leaning towards maintaining the same behavior between clang's braced-initialization and parenthesized initialization here, but if others have any other ideas, I'm happy to hear them.

Reply:
That's fine.